common/thorhelper/thorfile.cpp

```diff
@@ -100,3 +101,264 @@ IHThorDiskReadArg * createWorkUnitReadArg(const char * filename, IHThorWorkunitR
+    if ((uncompressedSize != 0) && (compressedSize != 0))
+        result.sizeOriginalData = uncompressedSize;
+
+    //The following will depend on the compression format - e.g. if compressed searching is implemented
+    result.sizeMemoryBranches = result.sizeOriginalBranches;
+
+    //NOTE: sizeOriginalData now includes the blob sizes that are removed before passing to the builder
+    //      if the original blob size is recorded then use it, otherwise estimate it
+    if (result.sizeOriginalData)
+    {
+        offset_t originalBlobSize = attrs.getPropInt64("@originalBlobSize");
+        if (result.numBlobNodes == 0)
+            result.sizeMemoryLeaves = result.sizeOriginalData;
+        else if (originalBlobSize)
+            result.sizeMemoryLeaves = result.sizeOriginalData - originalBlobSize;
+        else
+            result.sizeMemoryLeaves = (offset_t)((double)result.numLeafNodes * result.sizeOriginalData)/(result.numLeafNodes + result.numBlobNodes);
+    }
+}
+
+bool calculateDerivedIndexInformation(DerivedIndexInformation & result, IDistributedFile * file, bool force)
+{
+    if (!checkIndexMetaInformation(file, force))
+        return false;
+
+    IDistributedSuperFile * super = file->querySuperFile();
+    if (super)
+    {
+        Owned<IDistributedFileIterator> subfiles = super->getSubFileIterator(true);
+        bool first = true;
+        ForEach(*subfiles)
+        {
+            IDistributedFile & cur = subfiles->query();
+            if (!first)
+            {
+                DerivedIndexInformation nextInfo;
+                gatherDerivedIndexInformation(nextInfo, &cur);
+                mergeDerivedInformation(result, nextInfo);
+            }
+            else
+            {
+                gatherDerivedIndexInformation(result, &cur);
+                first = false;
+            }
+        }
+    }
+    else
+        gatherDerivedIndexInformation(result, file);
+
+    return true;
+}
+
+
+void mergeDerivedInformation(DerivedIndexInformation & result, const DerivedIndexInformation & other)
+{
+    if (!other.knownLeafCount)
+        result.knownLeafCount = false;
+
+    result.numLeafNodes += other.numLeafNodes;
+    result.numBlobNodes += other.numBlobNodes;
+    result.numBranchNodes += other.numBranchNodes;
+    result.sizeDiskLeaves += other.sizeDiskLeaves;
+    result.sizeDiskBlobs += other.sizeDiskBlobs;
+    result.sizeDiskBranches += other.sizeDiskBranches;
+
+    // These are always known/derived
+    result.sizeOriginalBranches += other.sizeOriginalBranches;
+    result.sizeMemoryBranches += other.sizeMemoryBranches;
+
+    //These may or may not be known
+    if (other.sizeOriginalData)
+        result.sizeOriginalData += other.sizeOriginalData;
+    else
+        result.sizeOriginalData = 0;
+    if (other.sizeMemoryLeaves)
+        result.sizeMemoryLeaves += other.sizeMemoryLeaves;
+    else
+        result.sizeMemoryLeaves = 0;
```

Comment:
why zero these out when other is zero ?

Reply:
because the value is unknown so the aggregated value can not be used for a calculation